json-path/src/main/java/com/jayway/jsonpath/spi/json/AbstractJsonProvider.java

```diff
@@ -198,14 +198,17 @@ public Object copy(Object obj) {
 
     @Override
     public Object getProperty(Object obj, Object key) {
-        if (isMap(obj))
-            return ((Map) obj).get(key.toString());
-        else {
-            List list = (List) obj;
-            int index = key instanceof Integer ? (Integer) key : Integer.parseInt(key.toString());
-            return list.get(index);
-        }
-
+    	if (isMap(obj)) {
+    		Object mapValue = getMapValue(obj, key.toString());
+    		if (mapValue.equals(UNDEFINED)) {
+    			return null;
+    		} else {
+    			return mapValue;
+    		}
+    	} else {
+    		int index = key instanceof Integer ? (Integer) key : Integer.parseInt(key.toString());
+    		return getArrayIndex(obj, index);
```

Comment:
In your next PR also change the last line in getProperty () to be getArrayIndex()